src/main/java/io/libp2p/discovery/mdns/impl/JmDNSImpl.java

```diff
@@ -386,21 +396,19 @@ public void stop() {
         // close socket
         this.closeMulticastSocket();
 
-        while(true) {
-            boolean allDone = _executor.isShutdown();
-
-            for (Future<Void> f : shutdowns)
-                allDone &= f.isDone();
-
-            if (allDone)
-                break;
+        logger.debug("JmDNS waiting for service stop...");
 
+        for (Future<Void> shutdown : shutdowns) {
             try {
-                TimeUnit.MILLISECONDS.sleep(100);
-            } catch(InterruptedException e) { }
+                shutdown.get(10, TimeUnit.SECONDS);
+            } catch (CancellationException e) {
+                logger.trace("Task was already cancelled");
+            } catch (InterruptedException | ExecutionException | TimeoutException e) {
+                logger.debug("Exception when stopping JmDNS: ", e);
+                throw new RuntimeException(e);
```

Comment:
Don't we actually need to rethrow the raw `InterruptedException`?

Reply:
The canonical `InterruptedException` handling is just calling `Thread.currentThread().interrupt();`. Fixed this